test/com/sun/jna/MemoryTest.java

```diff
@@ -199,6 +204,43 @@ public void testRemoveAllocatedMemory() {
         assertEquals(0, Memory.integrityCheck());
     }
 
+    public void testGetSharedMemory() {
+        Memory mem1 = new Memory(512);
+        Memory mem2 = new Memory(512);
+
+        // Get pointers into the two memory objects
+        Pointer stringStart1 = mem1.share(128);
+        Pointer stringStart2 = mem2.share(128);
+
+        mem1.setPointer(10 * POINTER_SIZE, stringStart1);
+        mem1.setPointer(11 * POINTER_SIZE, stringStart2);
+
+        // The pointer in mem1 at offset 10 * POINTER_SIZE points into the
+        // memory region of mem1, while the pointer at offset 11 * POINTER_SIZE
+        // points to the second memory region
+
+        // It is expected, that resolution of the first pointer results in
+        // an instance of SharedMemory (a subclass of Memory, that retains a
+        // reference on the originating Memory object)
+        Assert.assertThat(mem1.getPointer(10 * POINTER_SIZE), instanceOf(Pointer.class));
+        Assert.assertThat(mem1.getPointer(10 * POINTER_SIZE), instanceOf(Memory.class));
+        // The second pointer lies outside of memory 1, so it must not be a
+        // Memory object, but a raw pointer
+        Assert.assertThat(mem1.getPointer(11 * POINTER_SIZE), instanceOf(Pointer.class));
+        Assert.assertThat(mem1.getPointer(11 * POINTER_SIZE), not(instanceOf(Memory.class)));
+
+        // It is expected, that Memory#read called for pointers shows the same
+        // behaviour as direct calls to getPointer calls with the corresponding
+        // offsets
+        Pointer[] pointers = new Pointer[2];
+        mem1.read(10 * POINTER_SIZE, pointers, 0, 2);
+
+        Assert.assertThat(pointers[0], instanceOf(Pointer.class));
+        Assert.assertThat(pointers[0], instanceOf(Memory.class));
+        Assert.assertThat(pointers[1], instanceOf(Pointer.class));
+        Assert.assertThat(pointers[1], not(instanceOf(Memory.class)));
+    }
```

Comment:
Consider adding a test for the expected `IndexOutOfBoundsException` on the new `read()` code.

Reply:
Not sure if the tests really bring much to the table, but then it was interesting to implement them.